motionplan/ik/metrics.go

```diff
@@ -187,9 +236,46 @@ func NewSquaredNormSegmentMetric(orientationScaleFactor float64) SegmentMetric {
 // SquaredNormNoOrientSegmentMetric is a metric which will return the cartesian distance between the two positions.
 func SquaredNormNoOrientSegmentMetric(segment *Segment) float64 {
 	delta := spatial.PoseDelta(segment.StartPosition, segment.EndPosition)
-	// Increase weight for orientation since it's a small number
 	return delta.Point().Norm2()
 }
 
+// WeightedSquaredNormSegmentMetric is a distance function between two poses to be used for gradient descent.
+// This changes the magnitude of the position delta used to be smaller and avoid numeric instability issues that happens with large floats.
+// It also scales the orientation distance to give more weight to it.
+func WeightedSquaredNormSegmentMetric(segment *Segment) float64 {
+	// Increase weight for orientation since it's a small number
+	orientDelta := spatial.QuatToR3AA(spatial.OrientationBetween(
+		segment.EndPosition.Orientation(),
+		segment.StartPosition.Orientation(),
+	).Quaternion()).Mul(orientationDistanceScaling).Norm2()
+	// Also, we multiply delta.Point() by 0.1, effectively measuring in cm rather than mm.
+	ptDelta := segment.EndPosition.Point().Mul(0.1).Sub(segment.StartPosition.Point().Mul(0.1)).Norm2()
+	return ptDelta + orientDelta
+}
+
 // TODO(RSDK-2557): Writing a PenetrationDepthMetric will allow cbirrt to path along the sides of obstacles rather than terminating
 // the RRT tree when an obstacle is hit
+
+// FSConfigurationDistance is a fs metric which will sum the abs differences in each input from start to end.
+func FSConfigurationDistance(segment *SegmentFS) float64 {
+	score := 0.
+	for frame, cfg := range segment.StartConfiguration {
+		if endCfg, ok := segment.EndConfiguration[frame]; ok && len(cfg) == len(endCfg) {
```

Comment:
[q]: should the case where ok is not true be handled here?

Reply:
In the case where ok is not true, it is properly "handled" by contributing zero towards the score, as the frame would be in the same configuration at both ends of the segment.